File: src/main/java/io/mfedirko/contactme/ContactMeController.java
package io.mfedirko.contactme;

import jakarta.validation.Valid;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Controller;
import org.springframework.ui.ModelMap;
import org.springframework.validation.Errors;
import org.springframework.web.bind.annotation.*;

@Controller
@RequestMapping("/contact")
@RequiredArgsConstructor
@Slf4j
public class ContactMeController {
    private static final String CONTACTME_FORM = "contactme";
    private static final String CONTACTME_RECEIVED = "contactme-received";

    private final ContactMeRepository repository;

    @GetMapping
    public String getPage(ModelMap modelMap) {
        modelMap.addAttribute("contactForm", new ContactForm("", "", ""));
        return CONTACTME_FORM;
    }

    @PostMapping
    public String submitForm(@Valid ContactForm form, Errors errors, ModelMap modelMap) {
        if (errors.hasErrors()) {
            return CONTACTME_FORM;
        }

        log.info("Received contact request from {}<{}>", form.getFullName(), form.getEmail());
        repository.save(form);
        modelMap.addAttribute("name", form.getFullName());
        return CONTACTME_RECEIVED;
    }
}

File: src/main/kotlin/io/mfedirko/contactme/ContactMeController.kt
package io.mfedirko.contactme

import io.mfedirko.common.util.logger
import jakarta.validation.Valid
import org.springframework.stereotype.Controller
import org.springframework.ui.ModelMap
import org.springframework.validation.Errors
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestMapping

@Controller
@RequestMapping("/contact")
class ContactMeController(
    private val repository: ContactMeRepository
){
    private val log = logger()

    @GetMapping
    fun getPage(modelMap: ModelMap): String {
        modelMap.addAttribute("contactForm", ContactForm("", "", "", ""))
        return CONTACTME_FORM
    }

    @PostMapping
    fun submitForm(@Valid form: ContactForm?, errors: Errors, modelMap: ModelMap): String {
        if (errors.hasErrors()) {
            return CONTACTME_FORM
        }
        log.info("Received contact request from {}<{}>", form!!.fullName, form.email)
        repository.save(form)
        modelMap.addAttribute("name", form.fullName)
        return CONTACTME_RECEIVED
    }

    companion object {
        private const val CONTACTME_FORM = "contactme"
        private const val CONTACTME_RECEIVED = "contactme-received"
    }
}